### ISSUE TITLE
Discussion about Forge Update Checker

### ISSUE THREAD

**Issue by thiakil** [MinecraftForge/MinecraftForge#4376]
I've seen various bits and pieces around about the [FUC](https://mcforge.readthedocs.io/en/latest/gettingstarted/autoupdate/) not being 'adequate' or not doing X or Y and even had some ideas myself about potential uses, so wanted to draw everything together so we can see what is/isn't possible/wanted, other things people do in their own checkers, etc. If I've missed any existing issues please let me know and I'll add to the list.

Things already ruled out:

- Auto download of updates (#2816)

Concerns I've seen:

- Modders using non-semver wanting to ensure their versions are parsed correctly, or special comparisons (though the latter seemed to be misunderstanding latest vs recommended)
    - ~~Perhaps a custom parser can be specified in the `@Mod`?~~ *Nope, semver only*
- Only used from `@Mod`, not `mcmod.info` (#3053)
    - I don't see this as much of an issue, however for my future-thoughts it would be useful to *recommend* it be added to the info file (on the FUC doc, already present [here](https://mcforge.readthedocs.io/en/latest/gettingstarted/structuring/)) so it can be easily found outside Java

My thoughts/concerns:

* Only changelog data is supported, without adding an extra top-level property.
```
"<mcversion>": {
    "<modversion>": "<changelog for this version>", 
```

I would suggest a backward compatible addition of using an object instead of a string

```
"<mcversion>": {
    "<modversion>": {
        "changelog": "<changelog for this version>", 
        "<custom prop>": "<custom val>"
    }
````

Following the above, we could add some standardised extra props like `downloadURL`, `dependencies` (`modid`s? links to update jsons? perhaps otherwise better for mcmod.info?)

Reasoning: useful for non-curse update checkers. If we can standardise the info, anyone can make a compatible system.

* Is it possible to access the parsed info from inside a mod's code?
* Should there be a subscribable event for an outdated mod?
    * The above 2 would allow for modders to do their obnoxious update notifications (so I'm hesitant), but can be disabled globally in forge config
* Is there a way for a mod to gather a list of *all* available updates?
    * Allows for a mod to add a separate GUI/notification of updates etc
* Potential other standard fields (like those available in `mcmod.info`, but that requires download of the mod, rather than say, an index of update jsons)
    * or a link to a generic `mcmod.info` without the version info (suboptimal due to extra request)

EDIT: I'm more concerned about the JSON format that gets published on a webserver. Other stuff included for discussion

**Comment by williewillus** [comment]
> Is it possible to access the parsed info from inside a mod's code?

This is already possible.

> Is there a way for a mod to gather a list of all available updates?

Loop through the loaded mod list and use the above mentioned method

**Comment by williewillus** [comment]
By possible, i mean accessing the outdated status of the mod currently

As for the extra properties, depends on if there's demand for it. Currently launchers handle this all using an external resource (curse)

**Comment by thiakil** [comment]
> use the above mentioned method 

What is the method that was "mentioned"?

> Currently launchers handle this all using an external resource (curse)

That's what I think we should be providing so it's not dependant on this or that third party for incompatible apis (doesn't necessarily need to be accessible in game-api though)

**Comment by williewillus** [comment]
I just meant that the answer to "gathering a list of all available updates" is just "use the method to get an individual mod's updates, and call it in a loop for all loaded mods"

**Comment by thiakil** [comment]
I am aware, but I'm asking what that method is; all you said about it was "This is already possible."

**Comment by LexManos** [comment]
>- Auto download of updates (#2816)

Yes, this will never happen. I am sorry but I will never allow random servers to run executable code on your client. That's just insane. Yes there are checks we can put into place, but no. Just take my 20+ years of experience in this field and just accept it.

>- Modders using non-semver wanting to ensure their versions are parsed correctly, or special comparisons (though the latter seemed to be misunderstanding latest vs recommended)
    - Perhaps a custom parser can be specified in the `@Mod`?


It was explicitly decided as I designed this system to not allow this. Modders need to follow standards. 1<2 Simply how it is. Deal with it. There will be holy wars on this, no. Flat out. 

>- Only used from `@Mod`, not `mcmod.info` (#3053)
    - I don't see this as much of an issue, however for my future-thoughts it would be useful to *recommend* it be added to the info file (on the FUC doc, already present [here](https://mcforge.readthedocs.io/en/latest/gettingstarted/structuring/)) so it can be easily found outside Java

Ya, people should add it to their mcmod.info its existed in there for quite some time, but never really formalized. I really wanna get some post processor into FG that will read your @Mod and build a mcmod.info file from it... Anyone wanna work on that?

>* Only changelog data is supported, without adding an extra top-level property.

Yes this is intentional, this is to keep it super simple and to prevent shit like what you just suggested. Adding download links or things like that that encourage the automation of the downloading. You get a website to link people to, use it.

>Reasoning: useful for non-curse update checkers. If we can standardise the info, anyone can make a compatibleit l system.

The info is standardized, it also has nothing to do with Curse. I proposed this to Curse ages before I posted the specs. But, as they typically do, didn't do anything with it so I had to go ahead with what I designed.

>* Is it possible to access the parsed info from inside a mod's code?

Probably, but honestly I don't think modders SHOULD ever touch this info. This is just encouraging nagging bullshit and popups saying "update your stuff!" The point is to be opt-outable, and non-intrusive.

>* Should there be a subscribable event for an outdated mod?

See above
>    * The above 2 would allow for modders to do their obnoxious update notifications (so I'm hesitant), but can be disabled globally in forge config

Oh hey you saw the obvious ;)

>* Is there a way for a mod to gather a list of *all* available updates?
    * Allows for a mod to add a separate GUI/notification of updates etc

Again, see above.

>* Potential other standard fields (like those available in `mcmod.info`, but that requires download of the mod, rather than say, an index of update jsons)
    * or a link to a generic `mcmod.info` without the version info (suboptimal due to extra request)

What other fields do you need? You have everything already in the currently existing mod...

**Comment by thiakil** [comment]
> It was explicitly decided as I designed this system to not allow this. Modders need to follow standards. 1<2 Simply how it is. Deal with it. There will be holy wars on this, no. Flat out.

Only suggested because some people don't like semver or already use another. I personally use semver or semver-like. I was also not aware of any discussion on such standardisation.

> Adding download links or things like that that encourage the automation of the downloading.

>The info is standardized, it also has nothing to do with Curse. I proposed this to Curse ages before I posted the specs. But, as they typically do, didn't do anything with it so I had to go ahead with what I designed.

>What other fields do you need? You have everything already in the currently existing mod...

Though I was not explicit, my aim *is* to enable easier automation in a standard place, but *not* inside Forge/Minecraft code. The JSON is completely external to the running code, there is no need for Forge to load anything but the changelog and version numbers.

Yes, allowing modders access to the data inside a running client can lead to more of the same, but what about a *separate* mod that people install that for example adds a box to the main menu that lists available updates without an extra click? It's pretty much an edge case, but doesn't need any more info than is already loaded. (also why it's a relatively smaller part of my post).

TLDR; I'm more focused on the JSON itself, other stuff mainly included for completeness.

**Comment by LexManos** [comment]
>Only suggested because some people don't like semver or already use another. I personally use semver or semver-like. I was also not aware of any discussion on such standardisation.

Ya, really its only semver-like. All it needs to be is a sequence of numbers, split by a period that we can do basic integer comparison to. What you wnna do beyond that is up to you.

>Though I was not explicit, my aim is to enable easier automation in a standard place, but not inside Forge/Minecraft code. 

I am NOT a fan of that. Especially if its something that is defined/encouraged by Forge.
HOWEVER. I AM a fan of people following standards... such as the maven repo layout standard... People *could* add a maven repo/artifact info to their mcmod.info file. Nothing on our end is stopping them.

I'm not seeing anything that needs to change on our end to do anything with this.

>but what about a separate mod that people install that for example adds a box to the main menu that lists available updates without an extra click?

*Insert slippery slope argument here* Just no, we're not going to encourage people doing this. There are external mod pack managers. If you want to deal with nasty update notifications use those. The entire point of the Forge one was to be REALLY unintrusive.

**Comment by thiakil** [comment]
> I am NOT a fan of that. Especially if its something that is defined/encouraged by Forge.

Why? You only want to have something that can say "There's an update available, go find it somewhere. Use some other standard if you wanna do that automatically"? Just because something is in the JSON the gets downloaded doesn't mean Forge has to do anything with it.

> People could add a maven repo/artifact info to their mcmod.info file. 

1. You completely missed the point about not needing the mod already downloaded. Adding as a field to the update JSON, sure, still needs to be a standardised field or it is pointless.
2. If we go with maven, everything can be retrieved from there, rendering your json redundant.
3. Not everyone uses maven, and if they did, see point 2.

> Insert slippery slope argument here Just no, we're not going to encourage people doing this. There are external mod pack managers. If you want to deal with nasty update notifications use those. The entire point of the Forge one was to be REALLY unintrusive.

Did you miss the point about that stupid example already being possible?

> The entire point of the Forge one was to be REALLY unintrusive.

Adding fields to a JSON file doesn't change that unless YOU change what Forge's code does.

**Comment by LexManos** [comment]
Yo seem to fail to understand what defining a standard means. If we define the data needed to automate the downloading of mods people will expect and think we endorse the downloading of mods. There are a lot of things you CAN do. But they are far from things you SHOULD. And adding it to the standard is saying you SHOULD do them.

As for having update info from maven. That's actually an extension of the maven format that isn't really standardized due to versioning issues. But that's a whole other topic.

But ya, all I'm getting here is that you're wanting us to edit the standard so that you can automate downloading and installing of mods (complaining you need the mcmod.info really...). This is a flat out no, never going to happen.

**Comment by thiakil** [comment]
All I'm getting here is that you don't want users downloading their mods in any non-manual way. Either that or you think adding it to spec means you *have* to add it into Forge. It really doesn't. A launcher would find the info useful though.

I mean, how do you think modpacks work? The vast majority of people don't search out and download their mods individually.

**Comment by LexManos** [comment]
I understand what you want exactly and why you want it. I understand that forge would not be downloading mods even if this was in the spec. Get me quite clear on this. I am not a moron. Again what I am saying is that by definition adding it to the spec officially means that we officially endorse people using it in that.mannor. Which Forge *will never* do. End of story. This is never going into the spec because it is a bad thing and Forge will never endorse people doing it. Yes I fully understand modnpacks exist and people use them. But that is a completely separate discussion. Topic closed because I've made myself quite clear many times and you are refusing to accept no as the answer.
